lib/LedgerSMB/Upgrade_Tests.pm

```diff
@@ -200,27 +214,69 @@ Enabled buttons
 
 subtype 'button'
     => as 'Str'
-    => where { $_ =~ /Save and Retry|Cancel|Force/ }
+    => where { $_ =~ /Save and Retry|Cancel|Force|Skip/ }
     => message { "Invalid button '$_'" };
 
 has buttons => (is => 'ro', isa => 'ArrayRef[button]',
-                default => sub {['Save and Retry', 'Cancel']},
-                required => 0);
+    default => sub { return ['Save and Retry', 'Cancel']}, required => 0);
 
 =item tooltips
 
-Tooltip for each button
+Tooltip for each button.
+Validate that buttons are enabled for each tooltip, then prepend defaults
+and override with test specific labeling.
 
 =cut
 
 has tooltips => (is => 'ro',
-    isa => 'HashRef[Str]',
-    default => sub {
-        return {
-            'Save and Retry' => marktext('Save the fixes provided and attempt to continue migration'),
-            'Cancel' => marktext('Cancel the <b>migration</b>')
-    }},
-    required => 0);
+    isa => 'Maybe[HashRef[Str]]', required => 0,
+    default => undef,   # Force initializer call
+    initializer => sub {
+        my ( $self, $value, $writer_sub_ref, $attribute_meta ) = @_;
+        $value //= {};
+        for my $btn (keys %$value) {
+            die "No button '$btn' in test '$self->{name}'"
+                if not grep( /^$btn$/, @{$self->{buttons}});
+        }
+        my %defaults = ('Save and Retry' => marktext('Save the fixes provided and attempt to continue migration'),
+                                'Cancel' => marktext('Cancel the <b>whole migration</b>'));
+        %$value = (%defaults,%$value);
+        $writer_sub_ref->($value);
+    }
+);
+
+=item skipable
+
+Can this test be skipped
+
+=cut
+
+has skipable => (is =>'ro', isa => 'Maybe[Bool]', lazy => 1,
+                 default =>  sub {
+                    return grep(/^Skip$/, @{$_[0]->{buttons}}) == 1;
+                 }
+);
+
+=back
+
+=head1 Validate the object
+
+=over
+
+=item BUILD
+
+Called after object building. Ideal to validate object fields for coherence
+
+=cut
+
+sub BUILD {
+    my $self = shift;
+    $self->_validate_displayed_key(keys %{$self->{selectable_values}})
```

Comment:
This check is done at run time (when the user is upgrading). However, it's validating information that's statically available in this file. For example the `no_business_for_customer` check defines all the components required to run this check. So, even without having a database to migrate, we can just check the column-consistency when running our `prove` tests.
And once we have established our `prove` outcome to be correct, why would we still want to run this check during a migration?

Reply:
Moved to `xt/50-upgrade-tests.t`. Same for correlation between tooltips and buttons.